multicluster/controllers/multicluster/leader/stale_controller.go

```diff
@@ -0,0 +1,234 @@
+			klog.ErrorS(err, "Failed to delete stale MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(&memberClusterAnnounce))
+			return
+		}
+	}
+}
+
+func (c *StaleResCleanupController) RunPeriodically(stopCh <-chan struct{}) {
+	klog.InfoS("Starting StaleResCleanupController")
+	defer klog.InfoS("Shutting down StaleResCleanupController")
+
+	ctx, _ := wait.ContextForChannel(stopCh)
+	go wait.UntilWithContext(ctx, c.cleanUpExpiredMemberClusterAnnounces, memberClusterAnnounceStaleTime/2)
+	<-stopCh
+}
+
+func (c *StaleResCleanupController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	memberAnnounce := &mcv1alpha1.MemberClusterAnnounce{}
+	err := c.Get(ctx, req.NamespacedName, memberAnnounce)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if err == nil && memberAnnounce.DeletionTimestamp.IsZero() {
+		// Ignore the event if it's not with non-zero DeletionTimestamp
+		return ctrl.Result{}, nil
+	}
+
+	// Clean up all corresponding ResourceExports when the member cluster's
+	// MemberClusterAnnounce is deleted.
+	clusterID := getClusterIDFromName(req.Name)
+	staleResExports, err := getResourceExportsByClusterIDFunc(c, ctx, clusterID)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports by ClusterID", "clusterID", clusterID)
+		return ctrl.Result{}, err
+	}
+	if !deleteResourceExports(ctx, c.Client, staleResExports) {
+		return ctrl.Result{}, fmt.Errorf("failed to clean up all stale ResourceExports for the member cluster %s, retry later", clusterID)
+	}
+
+	// When cleanup is done, remove the Finalizer of this MemberClusterAnnounce.
+	finalizer := fmt.Sprintf("%s/%s", MemberClusterAnnounceFinalizer, memberAnnounce.ClusterID)
+	memberAnnounce.Finalizers = common.RemoveStringFromSlice(memberAnnounce.Finalizers, finalizer)
+	if err := c.Update(context.TODO(), memberAnnounce); err != nil {
+		klog.ErrorS(err, "Failed to update MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(memberAnnounce))
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (c *StaleResCleanupController) SetupWithManager(mgr ctrl.Manager) error {
+	// Add an Indexer for ResourceExport, so it can be filtered by the ClusterID.
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &mcv1alpha1.ResourceExport{}, indexKey, func(rawObj client.Object) []string {
+		resExport := rawObj.(*mcv1alpha1.ResourceExport)
+		return []string{resExport.Spec.ClusterID}
+	}); err != nil {
+		klog.ErrorS(err, "Failed to create the index")
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&mcv1alpha1.MemberClusterAnnounce{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		}).
+		Complete(c)
+}
+
+func cleanUpStaleResourceExports(ctx context.Context, mgrClient client.Client) error {
```

Comment:
where is the func called?

Reply:
This should be a stale function during addressing comments. I will delete this. Thanks.

Reply:
Removed